src/mds/ScrubStack.cc

```diff
@@ -281,8 +280,12 @@ bool ScrubStack::validate_inode_auth(CInode *in)
   }
 }
 
-void ScrubStack::scrub_dir_inode(CInode *in, bool *added_children, bool *done)
+void ScrubStack::scrub_dir_inode(CInode *in, bool *added_children, bool *done, bool is_eval_stray)
 {
+  if(is_eval_stray) {
+    mdcache->maybe_eval_stray(in);
+    return;
+  }
```

Comment:
You didn't call the `on_finish` after evaling the stray. This is why it stuck inifinitly.

And this also seems incorrect, IMO you should eval the stray dentries instead of the `~mdsdir` or the `stray dirs`. So you need to iterate all the stray dirs and encqueue all the stray dentries under them. And then call the  `mdcache->maybe_eval_stray(in);` for each of them ?

Reply:
wont the recursive scrub traverse through all the stray dirs and enqueue them?

Reply:
this is now removed